### PR TITLE
[Combobox][Listbox.Option] Fixed click event bubbling causing Popover close

### PR DESCRIPTION
**Pull request by chloerice**
### WHY are these changes introduced?

It's a common pattern in the admin for `Combobox` to include a `Listbox.Action` to enable creation of new options. Because propagation of the click event on `Listbox.Option` is not stopped, when a `Listbox` is rendered in a `Popover`, like a `Combobox`, the `Popover` closes unexpectedly. `Combobox` currently has a hack to prevent this, but it's also a problem for searchable listbox implementations. 

### WHAT is this pull request doing?

This PR: 

- Stops propagation of the click event on `Listbox.Option` to prevent unwanted bubbling
- Removes the workaround for the bug in `Combobox`
- Fixes the `Autocomplete` `actionBefore` prop's `onAction` callback never being called when provided
- Adds the Listbox with search example and the Sheet/Popover with searchable listbox examples to Storybook
- Fixes the duplicate key React console error in the searchable listbox examples

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

- You can tophat `web` using [this Spinstance](https://shop1.shopify.test-listbox-bug-fix.chloe-rice.us.spin.dev/admin/products/1)
  OR
- Tophat the `Popover` "With searchable listbox" example in Storybook ([PR deploy](https://5d559397bae39100201eedc1-nefjqtdvzc.chromatic.com/?path=/story/all-components-popover--with-searchable-listbox), [Production example](https://polaris.shopify.com/components/popover)

I tested in `web` to ensure no regressions with this [Spinstance](https://shop1.shopify.test-listbox-bug-fix.chloe-rice.us.spin.dev/admin). There are no `Autocomplete` instances using `actionBefore`, which is probably why that bug was never caught. Tested several autocompletes, comboboxes, and listboxes, including the the "Browse all" action in the Product details OrganizationCard Type combobox. All work as expected and/or as in production. 

### 🎩 checklist

- [x] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [x] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [x] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Comment by chloerice** [comment]
Tested in web to ensure no regressions with this [Spinstance](https://shop1.shopify.test-listbox-bug-fix.chloe-rice.us.spin.dev/admin). There are no `Autocomplete` instances using `actionBefore`, which is probably why that bug was never caught. Tested several autocompletes, comboboxes, and listboxes, including the the "Browse all" action in the Product details OrganizationCard Type combobox and all work as expected and/or as in production.

File: polaris-react/src/components/Popover/Popover.stories.tsx [hunk 1]
```diff
@@ -221,3 +230,329 @@ export function WithLazyLoadedList() {
...
+      onActiveOptionChange={handleActiveOptionChange}
+    >
+      {segmentList}
+      {showAllMarkup}
+      {noResultsMarkup}
+      {lazyLoadingMarkup}
+    </Listbox>
+  );
+
+  console.log('visible option index', visibleOptionIndex);
```
Comment (brittcorry) [inline polaris-react/src/components/Popover/Popover.stories.tsx]: Is this log intentional? (Obviously not hurting anything in storybook but just thought I'd check!)

**Review by brittcorry — APPROVED** [review]
Wow @chloerice — quick work and nice overhaul! Everything looks to be working as expected to me and I don't see any issues with the code (other than one maybe unintentional log)!

Thank you so much for this fix 🙌

File: polaris-react/src/components/Popover/Popover.stories.tsx (hunk 1 above)
Comment (chloerice) [reply polaris-react/src/components/Popover/Popover.stories.tsx]: No leftover from tophatting, thank you!

**Pull request merged**